clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java

```diff
@@ -0,0 +1,97 @@
+import org.infinispan.client.hotrod.configuration.Configuration;
+import org.jboss.as.clustering.controller.ResourceServiceBuilder;
+import org.jboss.as.clustering.infinispan.InfinispanLogger;
+import org.jboss.as.controller.OperationContext;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.dmr.ModelNode;
+import org.jboss.msc.service.Service;
+import org.jboss.msc.service.ServiceBuilder;
+import org.jboss.msc.service.ServiceController;
+import org.jboss.msc.service.ServiceName;
+import org.jboss.msc.service.ServiceTarget;
+import org.wildfly.clustering.infinispan.spi.InfinispanRequirement;
+import org.wildfly.clustering.infinispan.spi.RemoteCacheContainer;
+import org.wildfly.clustering.service.InjectedValueDependency;
+import org.wildfly.clustering.service.SuppliedValueService;
+import org.wildfly.clustering.service.ValueDependency;
+
+/**
+ * @author Radoslav Husar
+ */
+public class RemoteCacheContainerBuilder implements ResourceServiceBuilder<RemoteCacheContainer> {
+
+    private ValueDependency<Configuration> configuration;
+    private final PathAddress address;
+    private final String name;
+
+    public RemoteCacheContainerBuilder(PathAddress address) {
+        this.address = address;
+        this.name = address.getLastElement().getValue();
+    }
+
+    @Override
+    public ServiceName getServiceName() {
+        return RemoteCacheContainerResourceDefinition.Capability.CONTAINER.getServiceName(this.address);
+    }
+
+    @Override
+    public RemoteCacheContainerBuilder configure(OperationContext context, ModelNode model) throws OperationFailedException {
+        this.configuration = new InjectedValueDependency<>(InfinispanRequirement.REMOTE_CONTAINER_CONFIGURATION.getServiceName(context, this.name), Configuration.class);
+        return this;
+    }
+
+    @Override
+    public ServiceBuilder<RemoteCacheContainer> build(ServiceTarget target) {
+        Function<RemoteCacheManager, RemoteCacheContainer> mapper = remoteCacheManager -> new ManagedRemoteCacheContainer(this.name, remoteCacheManager);
+
+        Supplier<RemoteCacheManager> supplier = () -> {
+            Configuration configuration = this.configuration.getValue();
+            RemoteCacheManager remoteCacheManager = new RemoteCacheManager(configuration);
+            remoteCacheManager.start();
+            InfinispanLogger.ROOT_LOGGER.remoteCacheContainerStarted(this.name);
+            return remoteCacheManager;
+        };
+
+        Consumer<RemoteCacheManager> destroyer = remoteCacheManager -> {
+            remoteCacheManager.stop();
+            InfinispanLogger.ROOT_LOGGER.remoteCacheContainerStopped(this.name);
+        };
+
+        Service<RemoteCacheContainer> service = new SuppliedValueService<>(mapper, supplier, destroyer);
+        ServiceBuilder<RemoteCacheContainer> builder = target.addService(this.getServiceName(), service)
```

Comment:
This needs to be wrapped using AsynchronousServiceBuilder(...) as the service start/stop methods are blocking.

Reply:
Fixed in https://github.com/wildfly/wildfly/pull/9355